### ISSUE TITLE
Basecamp 3 API?

### ISSUE THREAD

**Issue by archon810** [basecamp/bcx-api#196]
Hi,

Basecamp 3 looks great. I'd like to upgrade, but we have some integration with HipChat and BC going, and I have to make sure it works with BC3 first.

1. Is there an API for BC3?
2. Is there maybe existing integration planned with HipChat?

Thanks.

**Comment by jeremy** [comment]
Basecamp 3 will offer a brand new API, too. Coming soon! Will keep this open until we have a link to share 😊

**Comment by emersonveenstra** [comment]
Any update on this? We really want to move to BC3, but we use the API for a few custom integrations and until the API comes out, we can't switch

**Comment by nfm** [comment]
We're also interested in this at Paydirt to do our time tracking integration.

**Comment by cagfn** [comment]
Urgently supporting this request! We have existing integrations that we'd love to use with BC3.

**Comment by emersonveenstra** [comment]
Doing some digging, looks like there's hidden endpoints available ;) Not going to say them here, but hopefully they'll polish them up soon and release it

**Comment by EnLabWalt** [comment]
guys https://www.youtube.com/watch?v=uMyCa35_mOg

**Comment by smythp** [comment]
Any news on this? Would also like to support a request for a Basecamp 3 API.

**Comment by jeremy** [comment]
News: We'll post here as soon as we have an early API to share with you! 😊

**Comment by smythp** [comment]
Great! Thank you.

**Comment by rvanbaalen** [comment]
With all do respect this release schema you're following is killing your 'integration partners'. Users already have the latest and greatest Basecamp (v3) available but developers aren't able to update their integrations due to missing API documentation.

It has been a couple of months now and still not a single sign of API docs being released. Could you please provide us with a heads up on the ETA?

Some of my users state that they won't upgrade to Basecamp v3 until we're ready. That's nice of them. Others are leaving our paid plans due to lacking (and slow) Basecamp v3 support. Can't blame them.

Is there anything we could do to help speed things up?

**Comment by ubergeekzone** [comment]
What's  the status of the basecamp 3 API? i love the new basecamp but it lacks a few features that i require... such as tracking git commits to bitbucket and slack chat notifications.

**Comment by bestwnh** [comment]
Hope for some status update...

**Comment by cnhx** [comment]
Any update on this?

**Comment by t1mmen** [comment]
+1 to @rvanbaalen's comment. An ETA would be much appreciated.

**Comment by rvanbaalen** [comment]
@jeremy When will that 'early' (*late*) API become available? It sounds like you don't have a v3 API at all yet?

**Comment by xiamx** [comment]
+1 to @rvanbaalen , it would be nice to be more transparent on this.

**Comment by samuelsimoes** [comment]
Also waiting it to support new API on the Chrome Basecamp Notifier. :disappointed:

**Comment by eileencodes** [comment]
Hey everyone, we're definitely going to have an API but we don't have a timeline. I'm sorry that we can't give you a specific date, but know that an API is in the works. Thanks for your patience!

**Comment by bestwnh** [comment]
@eileencodes I understand. But my leader won't let me set up a todo with no due date... So at least a rough date like 3 months or something?

**Comment by raulpopadineti** [comment]
+1 to @rvanbaalen. We're still waiting for `the soon-to-be-released Basecamp 3 API`. Still no ETA?

**Comment by archondigital** [comment]
Any news on the brand new API?

**Comment by jeremy** [comment]
Thanks for the interest, all! No fixed ETA, but it's in the works. We'll update here the moment we have news to share.

**Comment by rvanbaalen** [comment]
With all do respect. Its nog just interest. Some of us have business critical processes linked the API that you guys decided to break with unannounced. I would have expected more developer love from Basecamp. 

---
This email was sent from my iPhone and therefore subject to typos and other inaccuracies.

**Comment by tomdidom** [comment]
+1 to @rvanbaalen.

**Comment by Fustrate** [comment]
Is there anyone *actually* working on the API, or is it just on a roadmap tucked away in a corner? It's been almost 3 months since v3 was released, and I hope you guys would've already been working on a public API before then.

> Thanks for your patience!

There's really no patience involved when we have no other option but to wait here in the dark.

**Comment by cagfn** [comment]
Part of the disconnect, I think, is that this is not just normal reaction
to the Basecamp "we don't announce future stuff" thing. I can imagine being
on the Basecamp team, reading this thread, and thinking "this is just part
of being on the team of an edgy company that does stuff our own way." I
think in this case, though, in 2016, an API is not just a feature that's
nice for people to have, it's a necessary part of any product like Basecamp
at launch. If you encourage people who use your product to build things
that rely on it, you have a responsibility to them. That responsibility
goes beyond, say, the responsibility you have to people asking you to add
in whatever niche thing they want into Basecamp. I think we'd all
appreciate more detailed information, and an acknowledgment that a basic
API is uniquely *not applicable* to the Basecamp philosophy of "we don't
have a timeline."

**Comment by jeremy** [comment]
I hear you! And totally agree. We'll discuss how to open up our progress to you sooner than later.

Basecamp 3 is designed with API integration in mind from the get-go. We're proving that out with our mobile apps.

Those of you blocked on API availability: please [email me](jeremy@basecamp.com) to share how'd you aim to integrate with Basecamp 3. It'll help guide our choices for the initial API release.

**Comment by rvanbaalen** [comment]
Lets not close up again and learn from earlier comments in this thread Jeremy. Why not discuss these things publicly here, on Github?

Regarding your question on how I would aim to integrate with Basecamp 3: take a look at the long list of applications that are compatible with Basecamp 2. How about making the basic v2 API compatible with v3? As in; getting a list of Basecamps os nothing new right? Or at least providing a way for developers to differentiate on Basecamp Classic (basecamp), Basecamp v2 (bcx) and Basecamp 3 (bc3?) accounts would be helpful. That way our applications wont have to randomly crash because a v3 user tries to connect on a v2 API. Right now, v3 users have the v2 identifier 'bcx' which is VERY unwanted. 

That could at least be a quick fix for us developers to provide decent feedback "Sorry Basecamp v3 isn't compatible with anything yet." to our users. 

---
This email was sent from my iPhone and therefore subject to typos and other inaccuracies.

**Comment by kennebel** [comment]
The company I work for is also stuck. We have been encouraged by Basecamp to convert "lock, stock, and barrel" over to BC3, but we have multiple processes that rely on third-party and internal integration. We cannot suddenly abandon our work flow.

I am fairly concerned with the idea that 2 hours ago, it was suggested that this is a great time to provide input on the direction of the API. Several months after release of the product, the API should only be delayed for bug testing, "guide our choices" should be so far in the rear-view mirror, that you can't pick it out anymore.

:(

**Comment by jeremy** [comment]
> Right now, v3 users have the v2 identifier 'bcx' which is VERY unwanted.

Definitely shouldn't be seeing this, @rvanbaalen. Send an email to support@basecamp.com and we'll troubleshoot with you. (I'm happy to discuss here, too.)

**Comment by jeremy** [comment]
@rvanbaalen, @kennebel, and all – I ask for feedback because I know you're sitting on your hands, waiting for release, feeling between a rock & hard place with your own customers and teams who depend on you. I'm here to help. Your feedback can clarify our release priorities: what to expose first vs. what may come later. *Squints in the rear-view mirror.* 😯

@kennebel, I'd love to hear about the workflow you've built with 3rd-party and in-house integrations. Up for sharing?

**Comment by ubergeekzone** [comment]
@jeremy I tried to reach you guy's through twitter about offering a helping hand with development to speed up your API. Shoot me an email at amy.bridges@invoate.com and we can chat a little more there about lending a helping hand.

**Comment by kennebel** [comment]
@jeremy We take all of the useful date information from basecamp, the budgeted hours from another service that tracks time budgets and per task time spent, and build a continuously updated chart that shows the next several months of anticipated work load. Helps with scheduling people, contractor resources, and the like. Rather than guessing "how busy" a team is, I can point at a column chart, and tell them exactly when they'll be working over time, and when they should take a day off. :)

**Comment by cbaci** [comment]
Oh .... I've got a dashboard inside me that wants to come out .... #musthaveapi ... please BC3 dev ... before it's too late ... #agencyprincipalsneedprettycharts

**Comment by acurley** [comment]
I'm super disappointed in Basecamp here.  I recommended this product to my wife and her company just assuming that Basecamp would have an API with which she could integrate I cannot believe that the company I believed to be one of the great homes of RoR and @dhh would neglect to put API integrations first and foremost in their development cycle.

**Comment by dhh** [comment]
We will get an API out. But we've had other, more important things higher
on the todo list after launch. If having an API available right now is the
most important thing about Basecamp for you, you can also give Basecamp 2 a
shot. Still a great product. If not, hopefully it won't be too much longer
before we have the API ready.

**Comment by troy** [comment]
@jeremy in answer to your last comment, the API target I'd use is "all incomplete todos I'm assigned" or if it's materially easier, "all incomplete todos I have access to." https://twitter.com/tacoapp/status/692520150088704001 has more.

**Comment by rvanbaalen** [comment]
@dhh With all do respect have you read the replies in this topic? 

Most of us aren't 'fiddling around with the Basecamp API'. We're all businesses who have hundreds (thousands) of users that rely on the connectivity between our apps and Basecamp. 

It's not up to us to choose which version our users pick. We have real-world users who are being pushed (marketed) by Basecamp to upgrade their accounts to the latest Basecamp (v3). That's all very understandable. The problem here relies in the fact that yet again a Basecamp employee confirms that there are 'more important things' than delivering an API in time, months after the Basecamp 3 release.

FYI; This topic started 5 november. The issue we're discussing here started the moment Basecamp 3 was live. For a company as big as Basecamp, I would expect that developing, testing and releasing an API with a new software release would take less than three months to accomplish. 

Regarding the prioritizing of Basecamp API features @jeremy asked about; please just make the v2 API compatible with v3 without adding whatever fancy new stuff you're planning so we can at least get things working the way they were working before and make our customers happy again.

**Comment by FrankRee** [comment]
we just started a few weeks ago with Basecamp 3, I saw all the integrations and got my team into it, now I had to realize that they are all working only with older Basecamp versions. Not so good.

So what should we do? Wait for the API or change back to the older versions? What would you recommend? Is there any idea about a release date for the API?

**Comment by troy** [comment]
@jeremy It seems like the lack of an API and integrations should be covered on https://basecamp.com/3/transition/faqs and very possibly on https://basecamp.com/transition.

**Comment by dhh** [comment]
Robin, I have. And I sympathize with the desire to get this done soon. I
want it done soon too. But while it may not seem like it, Basecamp isn't a
huge organization. We have some 12 developers who have to manage every
single version of Basecamp and other legacy products, as well as all new
development on Basecamp 3.

So priorities have to be made. We have lots of work left to do on Basecamp
3. Lots. The API is one of them. A migration path for people who want to
move Basecamp 2 stuff to Basecamp 3 is another. And the internal lists are
endless. So far, we've simply had more important things to tackle with the
crew that we have. Doesn't mean we don't want to do an API, just means that
we haven't yet.

Fair warning: The API will not be backwards compatible. Basecamp 3 is a
complete rewrite. Different features, different data model, different
everything. So the API is going to be different as well. (That's why
Basecamp 2 is still around and will be around until The End of The
Internet).

FrankRee, if there are must-have integrations you cannot do without, I'd
recommend using Basecamp 2. Even when we release the new API, it'll still
take a while for integrators to adopt it. And some may not.

**Comment by dhh** [comment]
That's a good point, Troy. I'll raise the issue and we'll add a note that
if you depend on Basecamp 2 integrations, you shouldn't upgrade yet. We
cover other features from earlier versions that aren't available on
https://basecamp.com/3/transition/should-we-go-bc3/. Featuring the API and
the integrations there makes sense. Thanks for the reminder!

**Comment by tomdidom** [comment]
Meanwhile on reddit: https://www.reddit.com/r/programming/comments/43jult/basecamp_once_one_of_the_greatest_saas_apps_now/

**Comment by acurley** [comment]
@dhh Thanks for the response.  I wish I had known a month ago what I know now. I would have recommended Basecamp 2 and only upgraded to 3 when it was ready. Too bad we can't go back from 3 to 2 and get the features and integrations we were banking on.

I look forward to using the v3 API when it is done.

**Comment by dhh** [comment]
I understand the frustration. But as some of the comments there proclaim: There really isn't much that I can say. You want the API released. I want the API released. The API will be released when it's done. We will update this thread when we have something to share.

**Comment by rvanbaalen** [comment]
@dhh Let's just hope that the urgency of releasing this API is very clear and it won't be tucked away.

If there's anything we can do to help (besides having patience), please let us know.

**Comment by CppStars** [comment]
It's so sad, people like @dhh representing BaseCamp make you realize Indie companies like BaseCamp are not going to be as reliable as the big companies when it comes to supporting real products and users.

**Comment by rvanbaalen** [comment]
I strongly disagree with @CppStars and comments like that don't add anything useful to this thread.

**Comment by sjmog** [comment]
@cppstars you do realise that @dhh, among other things, created Rails, right? It'll be ready when he says it's ready.

**Comment by roydejong** [comment]
@sjmog Oh, sorry, he built Rails? I suppose that means he's earned a free pass to keep his company's integration partners in the dark, then.

**Comment by rzhenghi** [comment]
I think @CppStars has a point.  We use Rails and have a lot of respect to @dhh too.  I won't worry about @CppStars' comments offending him.  If you haven't read basecamp's book "Getting Real" https://basecamp.com/books/Getting%20Real.pdf, it's a very well written book.  One point in the book is "ignore feature requests" because the real important ones will keep coming back to you.  If there are no complaints, that means no one cares about the API and it will never be delivered.

**Comment by gashach** [comment]
i am a new user of basecamp, but i use many api's.  i am happy waiting for a better product, rather than a quick fix.

**Comment by qume** [comment]
These guys are stars and will learn from this and do better in future. 

Lets learn with them. 

This is a great case study of how well meaning companies can go wrong. I'm looking forward to a post-mortem from them in a few months so they can help us all improve.

**Comment by beems** [comment]
We excitedly signed up for Basecamp 3, only to be put on hold because there was no API yet. Here I sit, stuck. I'm not thrilled about paying for Basecamp 2 AND Basecamp 3 if I can't use Basecamp 3 (which I can't until I have at least a basic API). All I really need in the API is the ability to add a to-do to a list and assign it to a specific user. If I could do that programmatically in any way I could move to Basecamp 3.

**Comment by cbaci** [comment]
@dhh Thanks for continuing to parse this thread. Also, thanks for taking the time to do something right! As a new user (never on BC2) our agency will use BC3, pressure to satisfy integration needs will build, the most important needs will prioritize themselves and we'll build out on the APIs once they're released. 

BTW, will there be a "marketplace" for 3rd party API based SaaS offerings in the "Basecamp-verse"?

**Comment by dhh** [comment]
Yes, plan is to have a page listing all the integrations, just like we did
with BC2. So there'll be one repository for people to find 3rd party
offerings. Won't really be a marketplace per se, as we're not involved with
any billing or what have you.

We'll get it right and we'll get it done! We got some great ideas for
improvements, for example with WebHooks, so hopefully the API will be worth
the wait. Thanks for your patience!

**Comment by rvanbaalen** [comment]
Sounds great @dhh! Let's just hope the wait won't be much longer ;-)

---
This email was sent from my iPhone and therefore subject to typos and other inaccuracies.

**Comment by timbroder** [comment]
Thanks @dhh. Looking forward to it

**Comment by jgosier** [comment]
My team is eagerly anticipating the BC3 API, in the mean time we've had to start using Slack (which we avoided for as long as possible). Here's hoping BC3's API is ready soon

**Comment by timbroder** [comment]
@jeremy my 3 main use cases for the API:

1. Hack together [email-in](https://basecamp.com/help/2/guides/projects/email-in "email-in") which is [missing](https://twitter.com/timothybroder/status/696784457055993856 "missing") from v3. I imagine I'd either script this myself or use Zapier
2. [Sync ToDo items](http://www.timbroder.com/2015/10/how-to-sync-basecamp-todos-to-omnifocus-or-todoist-.html "Sync ToDo items") with my tool of choice
3. Integrate notifications with Slack

Thanks

**Comment by joshmassey** [comment]
Any update on this?  I signed my org up for BC3 thinking that the BC2 API was completely compatible.  I was shocked to hear this was not the case.  My team and I would appreciate some news in the near future.

**Comment by joshmassey** [comment]
Also, my team doesn't need all of the functionality.  We simply want the ability to pull information out via cURL.

**Comment by dhh** [comment]
Joshua, no additional updates in the last two weeks. We have someone assigned to do the work and a plan for implementation. We will update once there's something to dig into. As mentioned earlier, the API for BC3 will be a new API. Not compatible with BC2.

**Comment by kennebel** [comment]
@dhh https://basecamp.com/3/transition/should-we-go-bc3 still doesn't reflect your earlier comment about mentioning the lack of an API. This inclusion might help deflect future confusion or frustration as more people see your advertising for the brand new version.

I do feel that the comment "if you depend on Basecamp 2 integrations, you shouldn't upgrade yet" is an unrealistic expectation. In the company I work for, most of the team has never heard of an API, and don't understand why programming doesn't fix this like before, "it looks similar to BC2, why can't you do what you used to? I just want the best version BC has to offer, figure it out." For the others on here running a business, I doubt any of them could convince their integration clients not to use the prominently advertised, big happy mountain version that is on the front page.  "tail wagging the dog"

**Comment by dhh** [comment]
Jeremy, sorry about that. Just got it updated now, so the temporarily
missing API is listed on https://basecamp.com/3/transition/should-we-go-bc3/.
Looking forward to being able to remove that as a blocking reason!

**Comment by kennebel** [comment]
@dhh Thanks. more public info is good.

**Comment by GioLepori** [comment]
Hi guys! The company I'm working for, **Timeneye**, has just released a [time tracking integration](http://blog.timeneye.com/basecamp-3-time-tracking-integration/) for the new Basecamp 3. At the moment we are using our Chrome Widget injected inside Basecamp. It's a temporary solution but, still, we've managed to create a fully working integration. Hopefully, the APIs will be released soon!

**Comment by dhh** [comment]
You should all be hearing from @qrush shortly. He's working on the API now
and will update this thread when we very shortly have something for you all
to play with. Thanks again for the patience!

**Comment by rvanbaalen** [comment]
Great news! Can't wait to get started and bring our users the good news. 

---
This email was sent from my iPhone and therefore subject to typos and other inaccuracies.

**Comment by raulpopadineti** [comment]
@dhh how soon is "very shortly"?

And thank you for keeping us up-to-date with the process.

**Comment by dhh** [comment]
The work is underway. That's all I can say at this point. @qrush will
follow up with the specific progress when he has something to share.

**Comment by cbaci** [comment]
@qrush, perhaps you can give a quick hint on the authentication mechanism you'll be using, Oauth2?

**Comment by qrush** [comment]
@cbaci Yep, it'll be OAuth2 just like the BC2 API. More on our authentication scheme is here: https://github.com/basecamp/api/blob/master/sections/authentication.md

**Comment by clarkguru** [comment]
Looking forward to V3.

**Comment by cbaci** [comment]
TY @qrush

**Comment by kennebel** [comment]
Out of curiosity, will the new API be designed "corporate programmer friendly"? By this I mean, I am a corporate programmer, I am NOT on every project/basecamp in the company. However, when mgmt wants a report, they want all projects included (some of which they may not be included on, because there wasn't a reason to).

At this time, that means that I have to store the username/password of a couple of different users in order to gain access to all projects. (i don't feel good about storing the "owner" account info in a piece of code, just like I would never use Windows "Administrator" in code)

I'd like to see in the API, an option that if the account I'm using is an administrative one, it can see everything. Certainly, if the credentials used are not in the administrator group limit the view.

Thoughts?

**Comment by qrush** [comment]
@kennebel good question! You'd probably want to store the OAuth tokens that an owner signed in with and *not* their username/password. Just like you'd do for implementing with any other service - you wouldn't store someone's Twitter password (or even ask for it) - you'd go through their authentication flow and store the access tokens after.

**Comment by EMIjess** [comment]
Is it crazy to still be interested in the timeline on which we can expect a public API for basecamp3?

**Comment by stevish-com** [comment]
@EMIjess I feel that way too. It's all I can do not to come on here every day and say "Is it done yet, is it done yet?" I have a few tasks on my docket that are only waiting on the Basecamp3 API release, and I'm itching to get started on them (client is more patient than I am!)

**Comment by rvanbaalen** [comment]
So many basecamp users are still so unaware about the fact that no third party is able to work with Basecamp 3 (yet..). I think Basecamp could do a better job informing their users about this. Perhaps even a blogpost or something. 

November 2015 was the date Basecamp 3 was released if Im correct. In internet tech terms thats almost a decade ago and still no ETA for the API. 

---
This email was sent from my iPhone and therefore subject to typos and other inaccuracies.

**Comment by qrush** [comment]
Hi everyone. I'd appreciate a few minutes of your time.

## Sorry.

We could have done a better job of setting expectations about our API and calming the doubts in this thread. For that I apologize on behalf of Basecamp and our team of developers, designers, and more. We rely on our development community to bring great things into (and out of) Basecamp, and [it shows](https://basecamp.com/2/extras). Thanks for your patience and bearing with us as we've launched a new version of our product.

## Let's get started.

I'd like to announce that we're starting an **early access program** that will let you get into our API now, with a few notes:

1. We have a small subset of API endpoints ready for use, with a roadmap for what's next on deck
2. This is purely for those wanting to line up their integrations for launch once the API is public + give us feedback on how it's working
3. We don't have a specific release date yet for public launch

To apply for this early access program - just leave a :+1: [reaction](https://github.com/blog/2119-add-reactions-to-pull-requests-issues-and-comments) on this comment and I'll add you to our Basecamp 3 API documentation repo. (Please don't leave a _real_ comment and spam the thread more for access!)

## Any questions?

If you have specific questions about the Basecamp 3 API, I'd be happy to [answer them privately](mailto:nick@basecamp.com), or in the new documentation's repo issues. I'd especially like to talk if you are working on a app/service for public consumption, and you'd like to be listed on a page for integrations with Basecamp 3 eventually, [like we have for Basecamp 2](https://basecamp.com/2/extras).

For those who just need a fully documented, ready to use for all customers now API, we're still not ready for public launch yet. Please rest assured: we're on it. I'll update this thread once we're public.

:steam_locomotive: :bow:

**Comment by qrush** [comment]
OK, so apparently GitHub only shows the names of the first 10 people who have reacted with a given emoji. Oops.

So either pick a different emoji to react with, or just [email me](mailto:nick@basecamp.com) for access. :love_letter:

**Comment by muesliq** [comment]
Will the new API enable us to migrate users from BC1 and BC2 to BC3 in a way that they keep their accounts? Or are we stuck with legacy products "until the end of the internet"?

**Comment by dhh** [comment]
Helge,

Not sure I understand the question. But you will be able to create and read
content from BC3 using the API. You can use this in combination with the
APIs for Basecamp Classic and Basecamp 2 to copy data between the different
versions. And yes, you're welcome to use any version of Basecamp until the
end of the internet.

On Sat, Apr 2, 2016 at 10:16 AM, Helge Fahrnberger <notifications@github.com

**Comment by muesliq** [comment]
David, I'm trying to find out beyond doubt  if I will be able to use the API to migrate content **and users** from BC1 to BC3, in a way that (my customers') existing user accounts continue to work. 

(As - for some misterious reason -  you provide no migration logic between from previous versions to 3.)

**Comment by dhh** [comment]
Helge, it's not likely that we will offer an API for creating users as
these requests would require knowledge of someone's password. Basecamp
users all have a Basecamp ID that can give them access to any number of
accounts.

We are, however, going to investigate an official migration path later this
year.

On Mon, Apr 4, 2016 at 10:09 AM, Helge Fahrnberger <notifications@github.com

**Comment by muesliq** [comment]
David, I'm not asking to create new users if that was unclear. I of course just want to transfer user references from BC1 projects to BC3 projects.

Or put differently: 

Will I be able to able to use the API to write my own migration if I don't want to wait until "later this year", let alone submit myself to the outcome of that "investigation"?

**Comment by stevish-com** [comment]
I think there's no need to "migrate" users since a basecamp 1 user can
simply be added to a basecamp in basecamp 3. All the basecamp use the same
users as far as I can tell. In my menu, I see one v1 basecamp I have access
to, one v2 basecamp, and 2 v3 basecamp and can easily switch between them.

-Stephen

**Comment by dhh** [comment]
Basecamp shares identities between all versions, but while there's only one
identity for many accounts, there's still one user for each account.

Helge, we'll need to think more about how to support that use case. As it
stands now, the API is authenticated for one user. That user cannot
impersonate/post as someone else on the account. If it could, I could use
my API authentication to post as Jason Fried. I don't think he would like
that :D,

If you got each of the users you want to migrate to authenticate via your
system, though, that should work. Basically, building a migrator per-user,
per-Basecamp. That'd be cool.

**Comment by muesliq** [comment]
So I am stuck in Basecamp Classic until maybe (!) some time (!) there will be a migration routine. I'm close to migrate to Teamwork.com, despite having to make all those customer users sign up there again.. *frustrated*

**Comment by dhh** [comment]
Helge, sorry for the frustration. I wish I could wave a magic wand and have
all the features I'd like for Basecamp done yesterday. We'd surely have
bullet-proof migration from all previous versions of Basecamp! We've made a
commitment to exploring migration in the first half of this year, and we
will surely do that.

In the mean time, I can recommend using Basecamp 3 for brand new projects
and Basecamps and leave data for previous projects in the version of
Basecamp you had them. That's what we're doing at Basecamp ourselves. We
have data in all three versions of Basecamp. They're linked through the
account menu, so it's just a click away to jump between them.

That way you also don't disrupt any in-flight projects. We found that to be
helpful at Basecamp at with most customers we've talked to.

Fully understand the desire of "hey, we just want it all in one place". But
this is where we are at the moment.

**Comment by misterzik** [comment]
@qrush hey I would like access to the beta, if it's still open nickjrod at gmail.com

Thanks Nick,
- Nick

**Comment by michaeldtodd** [comment]
Where exactly can we change between basecamp versions? I don't see any menu that has it listed...

**Comment by ivanbozic** [comment]
@michaeldtodd, in Basecamp 2 there should be a persistent black bar at the top of the screen where you can switch between versions.

In Basecamp 3, if you click on the Basecamp icon in the top left-hand corner, there should be a _Switch Accounts_ area where you can do the same thing.

**Comment by joehanchoi** [comment]
Any ETA on these APIs going public yet?

**Comment by rvanbaalen** [comment]
Would love to know an ETA as well since our integration is receiving more and more Basecamp 3 users (who converted from Basecamp 2) and keep asking us why our tool wont work. 

Right now all I can give them is "Ask Basecamp". I would love to be able to answer something along the lines of "Be patient one more week/month and we will offer BC3 support."

---
This email was sent from my iPhone and therefore subject to typos and other inaccuracies.

**Comment by javan** [comment]
Quick update: We hit a few speed bumps along the way, but work on the BC3 API is progressing and we hope to release it publicly soon. We don't have an exact ETA, and I apologize for the delays.

You can request early access to the API by filling out this form:  https://basecamp.wufoo.com/forms/request-early-access-to-the-basecamp-3-api/

**Comment by kepatel** [comment]
We've requested early access but haven't gotten any response.  For me this is a business critical issue and would like to be able to eliminate some of the manual processes that are in place in our organization due to our transition from BC2 to BC3.

**Comment by rvanbaalen** [comment]
I am completely with @bcjarrett here. Other companies with similar is smaller development teams & budget can develop entire applications in the time we have been waiting for 'just' an API.

I too have to keep disappointing users that 'Basecamp 3 isn't supported yet' and 'We don't know when we can support it'.

**Comment by raulpopadineti** [comment]
Most certainly they're too focused on other things. They don't feel like it's such an important thing to invest since it's not affecting their business directly (only with a small decrease in reputation from us).

Until they find some time to get this API version out in the public, they'll still tell us "it's done when it's done". We can't do anything else than wait.

If someone would be dedicated 100% of the time on this (from their side), there's no way you can't estimate an ETA.

**Comment by javan** [comment]
Hi all, once again I'm sorry for the delay. We have not done a good job communicating our progress. 

Nick left Basecamp, and it's taken me a while to get up to speed with the project. The transition happened at a particularly busy time (company meetup, RailsConf, and [welcoming in a class of interns](https://m.signalvnoise.com/what-we-learned-while-hiring-the-basecamp-summer-internship-class-3fc019f85165)) so I wasn't able to really dig in until recently. You have my full attention now 😁.

One of the challenges we're still working through is how to handle file attachments with Basecamp 3's API. I'll explain.

In Basecamp 2, to [create a message with attachments using the API](https://github.com/basecamp/bcx-api/blob/master/sections/messages.md#attaching-files), you simply pass an array of `attachments` params along with the message params. Basecamp 2 displays all attachments below the content its associated with so the whole process is pretty straightforward. 

In Basecamp 3, files can be attached anywhere. You can drop a file in a message and it renders right where you dropped it. The experience is great and it enables users to compose much richer content than they could in Basecamp 2. Translating that experience to the API is where things get tricky. Instead of looking for `attachments` params, we need to scan the content for attachments, extracting the location and identifying attributes along the way. We don't expect API consumers to construct the full set of HTML it takes to render an attachment so we're working on a well defined, minimal HTML structure that will be seamlessly serialized on the way in and out.

We actually have two programmers working on the attachments API, and I'm cruising along to finish up everything else. I don't want to make promises I can't keep, but I suspect the API will be publicly available in 4-6 weeks.

Today I gave 75 more people beta access to the API. That's everyone in the backlog. If you're interested in early access, submit your request here: https://basecamp.wufoo.com/forms/request-early-access-to-the-basecamp-3-api/. I'll be granting access more frequently now.

Thanks for your time,
-Javan

**Comment by cbaci** [comment]
@javan, thanks for picking it up. Change being the only constant is difficult on timelines! As we build out our integration, we're finding that it'd be great to have some form of master json schema document to aid in building out persistence model. Is it too forward to ask if such a thing exists? Grepping through the .md files to build out our integrated db structure is cumbersome.

**Comment by garek007** [comment]
It sounds like we still don't have an API yet? I guess I'll apply for early access...

**Comment by rafo** [comment]
At least, they changed their communication behavior 👎 

@javan: "We have not done a good job communicating our progress"

**Comment by javan** [comment]
The Basecamp 3 API will officially launch on July 20. Thanks for hanging in there everyone!

**Comment by javan** [comment]
🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀
### 🚀 https://github.com/basecamp/bc3-api 🚀
🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀

**Comment by wraybowling** [comment]
I love that I was reading this thread today, and that the API came out today. I love a happy ending.

**Comment by jdsimcoe** [comment]
Woohoo!!! 🎉 🎉 🎉  https://www.producthunt.com/tech/basecamp-3-api
